File: src/relay/qnn/csi_op/fsmn.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*!
 * \file src/relay/qnn/op/fsmn.cc
 * \brief QNN fsmn operator.
 */
#include <tvm/relay/analysis.h>
#include <tvm/relay/op_attr_types.h>
#include <tvm/relay/qnn/attrs.h>

#include "../op/op_common.h"
#include "../utils.h"

namespace tvm {
namespace relay {
namespace qnn {
TVM_REGISTER_NODE_TYPE(QnnCSIFsmnAttrs);

bool QnnCSIFsmnRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
                   const TypeReporter& reporter) {
  CHECK_EQ(types.size(), 6);
  const auto* data = types[0].as<TensorTypeNode>();
  // const auto* l_filter = types[1].as<TensorTypeNode>();
  // const auto* r_filter = types[2].as<TensorTypeNode>();
  // const auto* frame_sequence = types[3].as<TensorTypeNode>();
  if (data == nullptr) return false;

  const QnnCSIFsmnAttrs* param = attrs.as<QnnCSIFsmnAttrs>();
  CHECK(param != nullptr);
  // const int l_order = param->l_order;
  // const int r_order = param->r_order;
  // const int l_stride = param->l_stride;
  // const int r_stride = param->r_stride;

  auto oshape = data->shape;
  // auto l_filter_shape = l_filter->shape;
  // auto r_filter_shape = r_filter->shape;
  // auto frame_sequence_shape = frame_sequence->shape;
  // tvm::tir::ExprDeepEqual expr_equal;
  // expr_equal(l_filter_shape[0], l_order);
  // expr_equal(l_filter_shape[1], oshape[1]);
  // expr_equal(r_filter_shape[0], r_order);
  // expr_equal(r_filter_shape[1], oshape[1]);
  // expr_equal(frame_sequence_shape[0],
  //                           r_order * r_stride + (l_order - 1) * l_stride + 1);
  // expr_equal(frame_sequence_shape[1], oshape[1]);

  reporter->Assign(types[5], TensorType(oshape, data->dtype));

  return true;
}

// QNN Relu operator.
Expr MakeQnnCSIFsmn(Expr frame, Expr l_filter, Expr r_filter, Expr frame_sequence,
                    Expr frame_counter, int l_order, int r_order, int l_stride, int r_stride,
                    int unavailable_frames, Array<Array<IndexExpr>> q_params, DataType out_dtype,
                    String layer_name) {
  auto attrs = make_object<QnnCSIFsmnAttrs>();
  attrs->l_order = std::move(l_order);
  attrs->r_order = std::move(r_order);
  attrs->l_stride = std::move(l_stride);
  attrs->r_stride = std::move(r_stride);
  attrs->unavailable_frames = std::move(unavailable_frames);
  attrs->q_params = std::move(q_params);
  attrs->out_dtype = std::move(out_dtype);
  attrs->layer_name = std::move(layer_name);
  static const Op& op = Op::Get("qnn.csi.fsmn");
  return Call(op, {frame, l_filter, r_filter, frame_sequence, frame_counter}, Attrs(attrs), {});
}

RELAY_REGISTER_OP("qnn.csi.fsmn")
    .describe(R"code(Feedforwward sequential memory network.

- **frame**: data is a 2D array of shape
            (1, length)

- **l_filter**: l_filter is a 2D array of shape
            (l_order, length)

- **r_filter**: r_filter is a 2D array of shape
            (l_order, length)

- **frame_sequence**:
            (l_order + r_order, length)

- **out**: Output is a 2D array of shape
            (1, length).

)code" TVM_ADD_FILELINE)
    .set_attrs_type<QnnCSIFsmnAttrs>()
    .set_num_inputs(5)
    .add_argument("data", "Tensor", "The input tensor")
    .add_argument("l_filter", "Tensor", "The parameter tensor")
    .add_argument("r_filter", "Tensor", "The parameter tensor")
    .add_argument("frame_counter", "Tensor", "count passed frames")
    .add_argument("frame_sequence", "Tensor", "Temporary space to hold past and future frames")
    .set_support_level(11)
    .add_type_rel("QnnCSIFsmnRel", QnnCSIFsmnRel)
    .set_attr<TOpPattern>("TOpPattern", kOpaque);

TVM_REGISTER_GLOBAL("relay.qnn.op._make.CSIFsmn").set_body_typed(MakeQnnCSIFsmn);

}  // namespace qnn
}  // namespace relay
}  // namespace tvm

File: python/tvm/topi/nn/fsmn.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
# pylint: disable=invalid-name, unused-variable, too-many-locals
# pylint: disable=unused-argument, redefined-builtin
"""fsmn operators"""
from .. import te, tir
from ..utils import get_const_tuple


def fsmn(
    frame,
    l_filter,
    r_filter,
    frame_sequence,
    frame_counter,
    l_order,
    r_order,
    l_stride,
    r_stride,
    unavailable_frames,
):
    """fsmn operator.

    Parameters
    ----------
    Input : tvm.te.Tensor
        2-D with shape [1, in_width]

    l_filter : tvm.te.Tensor
        2-D with shape [l_order, in_width]

    r_filter : tvm.te.Tensor
        2-D with shape [r_order, in_width]

    frame_sequence: tvm.te.Tensor
        2-D with shape [(l_order -1) * l_stride + r_order * r_stride, in_width ]

    l_stride : int
        Left stride size

    r_stride : int
        Right stride size

    Returns
    -------
    Output : tvm.te.Tensor
        2-D with shape [1, in_width]
    """

    def _fsmn(
        frame,
        l_filter_,
        r_filter_,
        frame_sequence,
        frame_counter,
        out_buf,
        l_order,
        r_order,
        l_stride,
        r_stride,
        unavailable_frames,
    ):
        ib = tir.ir_builder.create()
        last_frame = ib.buffer_ptr(frame)
        past_filter = ib.buffer_ptr(l_filter_)
        future_filter = ib.buffer_ptr(r_filter_)
        sequence_frame = ib.buffer_ptr(frame_sequence)
        frame_counter_ = ib.buffer_ptr(frame_counter)
        out = ib.buffer_ptr(out_buf)

        # len_order = (l_order -1) * l_stride + r_order * r_stride
        len_order, length = get_const_tuple(frame_sequence.shape)

        with ib.for_range(0, length) as i:
            out[i] = 0.0

        frame_counter_[0] = frame_counter_[0] + 1

        # set last frame to sequence tail.
        with ib.if_scope(frame_counter_[0] > unavailable_frames):
            with ib.for_range(0, len_order) as i:
                with ib.for_range(0, length) as j:
                    new_index = i * length + j
                    with ib.if_scope(i == (len_order - 1)):
                        sequence_frame[new_index] = last_frame[j]
                    with ib.else_scope():
                        original_index = (i + 1) * length + j
                        sequence_frame[new_index] = sequence_frame[original_index]

        # past frame
        with ib.for_range(0, l_order) as k:
            with ib.for_range(0, length) as l:
                in_index = k * l_stride * length + l
                filter_index = (l_order - k - 1) * length + l
                out[l] = past_filter[filter_index] * sequence_frame[in_index] + out[l]

        # current frame
        with ib.for_range(0, length) as m:
            in_index = (l_order - 1) * length * l_stride + m
            out[m] = sequence_frame[in_index] + out[m]

        # future frame
        with ib.for_range(0, r_order) as m:
            with ib.for_range(0, length) as n:
                in_index = m * r_stride * length + n + l_order * l_stride * length
                filter_index = m * length + n
                out[n] = future_filter[filter_index] * sequence_frame[in_index] + out[n]

        return ib.get()

    out_shape = list(get_const_tuple(frame.shape))
    out = te.extern(
        out_shape,
        [frame, l_filter, r_filter, frame_sequence, frame_counter],
        lambda ins, outs: _fsmn(
            ins[0],
            ins[1],
            ins[2],
            ins[3],
            ins[4],
            outs[0],
            l_order,
            r_order,
            l_stride,
            r_stride,
            unavailable_frames,
        ),
        dtype=frame.dtype,
    )

    return out
